packages/react-native/ReactCommon/react/performance/timeline/PerformanceObserver.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "PerformanceObserver.h"
+#include "PerformanceEntryReporter.h"
+
+namespace facebook::react {
+
+void PerformanceObserver::handleEntry(const PerformanceEntry& entry) {
+  // https://w3c.github.io/performance-timeline/#takerecords-method (step 7.1)
+  // https://www.w3.org/TR/event-timing/#should-add-performanceeventtiming
+  if (entry.entryType == PerformanceEntryType::EVENT) {
+    if (entry.duration < durationThreshold_) {
+      // The entries duration is lower than the desired reporting threshold,
+      // skip
+      return;
+    }
+  }
+
+  if (observedTypes_.contains(entry.entryType)) {
+    buffer_.push_back(entry);
+  }
+}
+
+std::vector<PerformanceEntry> PerformanceObserver::takeRecords() {
+  std::vector<PerformanceEntry> result;
+  buffer_.swap(result);
+  flush();
+  return result;
+}
+
+void PerformanceObserver::observe(
+    PerformanceEntryType type,
+    PerformanceObserverObserveSingleOptions options) {
+  // we assume that `type` was checked on JS side and is correct
+  observedTypes_.clear();
+  observedTypes_.insert(type);
+
+  durationThreshold_ = options.durationThreshold;
+
+  requiresDroppedEntries_ = true;
+
+  if (options.buffered) {
+    auto& reporter = PerformanceEntryReporter::getInstance();
+    buffer_ = reporter->getEntriesByType(type);
+
+    if (!buffer_.empty()) {
+      scheduleFlushBuffer();
+    }
+  }
+  registry_.addObserver(shared_from_this());
+}
+
+void PerformanceObserver::observe(
+    std::unordered_set<PerformanceEntryType> types,
+    PerformanceObserverObserveMultipleOptions options) {
+  observedTypes_ = std::move(types);
+  requiresDroppedEntries_ = false;
+  durationThreshold_ = options.durationThreshold;
+  registry_.addObserver(shared_from_this());
+}
+
+double PerformanceObserver::getDroppedEntriesCount() noexcept {
+  double droppedEntriesCount = 0;
+
+  if (requiresDroppedEntries_) {
```

Comment:
This should be a set of types that we need to report dropped entries. This should also work if we do this:

```
observer.observe({type: 'mark', buffered: true}); // report marks until this point and report dropped ones
observer.observer({type: measure', buffered: true}); // same, but for this time and until this moment
```

Reply:
We iterate over observer types below, and the flag is reset after each `observe`. So it should work as mentioned (considering `observe` resets entries, not adds more)

Reply:
I think we're not handling the timing of that correctly. We're determining the number of dropped entries when the callback is executed, not when we "dump" all the entries in the observer buffer. Also, if we execute the code I shared in previous example, we'd be dropping all the existing entries in the observer buffer because we just reassign it. We just append those and determine dropped entries at the time of observation.

Reply:
Hmm, I see your point now. But this behaves exactly as requested in spec https://w3c.github.io/performance-timeline/#observe-method

Reply:
You're absolutely right, and I think that's a mistake on the spec to be honest. You might be reporting a non-zero number of dropped entries when those entries are already in the observer's buffer. Doesn't make much sense, but ok to leave it like that. Thanks for checking!

Reply:
Filed https://github.com/w3c/performance-timeline/issues/214